include/gz/math/TimeVaryingVolumetricGridLookupField.hh

```diff
@@ -0,0 +1,231 @@
+        const T& time, const VolumetricGridLookupField<V> &_field) {}
+
+      /// \brief Creates a session for querying
+      public: S CreateSession() const {}
+
+      /// \brief Creates a session startingg at time T
+      /// \param
+      public: S CreateSession(const T &_time) const {}
+
+      /// \brief Update session to new time. Returns new session pointer if
+      /// more is available. Otherwise returns nullopt if we have run out of
+      /// time.
+      public: std::optional<S> StepTo(const S &_session, const T &_time) {}
+
+      /// \brief Looks up interpolators at a given time step. Use the session
+      /// to set the time step.
+      public: std::pair<InterpolationPoint4D<T, V>, InterpolationPoint4D<T, V>>
+      LookUp(const S &_session,
+        const Vector3<V> &_point,
+        const Vector3<V> &_tol) const {}
+
+      public: template<typename X>
+      X EstimateQuadrilinear(
+        const S &_session,
+        const std::pair<InterpolationPoint4D<T, V>,
+          InterpolationPoint4D<T, V>> &_points,
+        const std::vector<X>& values1,
+        const std::vector<X>& values2
+      ) const
+      {
+      }
+    };
+    /// \brief An in-memory session. Loads the whole dataset in memory and
+    /// performs queries.
+    template<typename T, typename V>
+    class InMemorySession
+    {
+      /// \brief Iterator which holds pointer to current state
+      /// TODO(arjo): Use friend to make visible only to InMemorySession
+      /// specialization
+      typename std::map<T, VolumetricGridLookupField<V>>::const_iterator iter;
+
+      /// \brief Time of last query
+      T time;
+
+      friend class
+        TimeVaryingVolumetricGridLookupField<T, V, InMemorySession<T, V>>;
+    };
+
+    /// \brief Specialized version of `TimeVaryingVolumetricGridLookupField`
+    /// for in-memory lookup. It loads the whole dataset into memory.
+    template<typename T, typename V>
+    class TimeVaryingVolumetricGridLookupField<T, V, InMemorySession<T, V>>
+    {
+      public: TimeVaryingVolumetricGridLookupField()
+      {}
+
+      public: void AddVolumetricGridField(
+        const T &_time, const VolumetricGridLookupField<V> &_field) {
+        this->gridFields.emplace(_time, _field);
+      }
+
+      public: InMemorySession<T, V> CreateSession() const {
+        InMemorySession<T, V> sess;
+        sess.iter = this->gridFields.begin();
+        sess.time = T(0);
+        return sess;
+      }
+
+      public: InMemorySession<T, V> CreateSession(const T &_time) const {
+        InMemorySession<T, V> sess;
+        sess.iter = this->gridFields.lower_bound(_time);
+        sess.time = _time;
+        return sess;
+      }
+
+      public: std::optional<InMemorySession<T, V>> StepTo(
+        const InMemorySession<T, V> &_session, const T &_time) const {
+        if (_session.iter == gridFields.end())
+        {
+          return std::nullopt;
+        }
+
+        InMemorySession<T, V> newSess(_session);
```

Comment:
@arjo129 nit: can we avoid a session copy if not necessary? User code will step frequently, as it can't know (nor should it have to know) when a step will actually have an effect.

Reply:
So something like `bool StepTo(InMemorySession<T, V> &_session, const T &_time) const` instead and modify the iterator in place?

Reply:
That's an option. The alternative would be to encapsulate all access through a mutable session (ie. cursor) and step the session (ie. cursor) itself.